### PR TITLE
Implement session functionality

### PR DESCRIPTION
**Pull request by tsione**
Add ability to pass session data over 'GraphQL' context to 'HTTP' request.

### CODE REVIEW

File: src/main/scala/core/graphql/UserContextFactory.scala [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package core.graphql
+
+import akka.http.scaladsl.model.HttpHeader
+import akka.http.scaladsl.model.headers.HttpCookie
+
+import scala.collection.mutable.ListBuffer
+
+class UserContextFactory {
```
Comment (tverdokhlebd) [inline src/main/scala/core/graphql/UserContextFactory.scala]: You can remove UserContextFactory class and use Companion pattern.

Comment (larixer) [reply src/main/scala/core/graphql/UserContextFactory.scala]: @tverdokhlebd In english please in the future

File: src/main/scala/core/controllers/graphql/HttpHandler.scala [hunk 2]
```diff
@@ -26,123 +29,118 @@ import scala.util.control.NonFatal
...
-                        ServerSentEvent(error.getMessage)
+  def handleQuery(graphQlMessage: GraphQLMessage): Route = extractRequest {
+    request =>
+      withHeaders(userContextFactory.createContext(request.headers.toList)) {
+        userCtx =>
+          QueryParser.parse(graphQlMessage.query) match {
+            case Success(queryAst) =>
+              val headers: ListBuffer[HttpHeader] = ListBuffer.empty
+              queryAst.operationType(graphQlMessage.operationName) match {
+                case Some(Subscription) =>
```
Comment (tverdokhlebd) [inline src/main/scala/core/controllers/graphql/HttpHandler.scala]: Subs must not use headers/cookies implementation.

File: src/main/scala/core/controllers/graphql/WebSocketHandler.scala [hunk 3]
```diff
@@ -20,9 +21,10 @@ import spray.json._
 import scala.util.{Failure, Success}
 
 @Singleton
-class WebSocketHandler @Inject()(graphQlExecutor: Executor[Unit, Unit])
+class WebSocketHandler @Inject()(graphQlExecutor: Executor[UserContext, Unit],
```
Comment (tverdokhlebd) [inline src/main/scala/core/controllers/graphql/WebSocketHandler.scala]: Websockets must not use headers/cookies implementation.

File: src/main/scala/common/models/Error.scala [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package common.models
+
+class Error extends Exception
+
+case class InternalServerError(msg: String = "") extends Error
```
Comment (tverdokhlebd) [inline src/main/scala/common/models/Error.scala]: There is this internal server error in the separate file.

File: src/main/scala/core/controllers/graphql/HttpHandler.scala [hunk 5]
```diff
@@ -26,123 +29,118 @@ import scala.util.control.NonFatal
...
-                operationName = graphQlMessage.operationName,
-                variables = graphQlMessage.variables.getOrElse(JsObject.empty)
-              ).map {
-                preparedQuery =>
-                  ToResponseMarshallable(preparedQuery.execute()
-                    .map(r => ServerSentEvent(r.compactPrint))
-                    .recover {
-                      case NonFatal(error) =>
-                        ServerSentEvent(error.getMessage)
+  def handleQuery(graphQlMessage: GraphQLMessage): Route = extractRequest {
```
Comment (tverdokhlebd) [inline src/main/scala/core/controllers/graphql/HttpHandler.scala]: You can move extracting request and mapping header/cookies to GraphQLController class.

**Review by tverdokhlebd — CHANGES_REQUESTED** [review]
See questions

**Review by tverdokhlebd — APPROVED** [review]

**Pull request merged**